csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,246 @@
+
+  // Returns if all atributes of the ID transforms first and second are the same
+  static bool transformAtributesMatch(Expr* first, Expr* second);
+
+  // Initializes entries for the provided IterDomain in the IterDomainGraphs
+  void initializeId(
+      IterDomain* id,
+      const VectorOfUniqueEntries<Expr*>& definitions,
+      const VectorOfUniqueEntries<Expr*>& uses);
+
+  // Returns if first and second are expressions through which the provided
+  // id_map have matching inputs (if forward), or outputs (if not forward).
+  // Returning true means the expressions are "the same", in terms they modify
+  // matching original extents, by the same amount.
+  bool exprsMap(
+      Expr* first,
+      Expr* second,
+      bool forward
+      // , std::vector<IterDomain*> second_input_or_output_override
+  ) const;
+
+  // Returns entry in unique_definitions_ for provided group in provided mode,
+  // otherwise errors if no entry is found.
+  ExprGroups uniqueDefinitions(IdGroup group) const;
+
+  // Returns entry in unique_uses_ for provided group in provided mode,
+  // otherwise errors if no entry is found.
+  ExprGroups uniqueUses(IdGroup group) const;
+
+  std::unordered_map<IdGroup, ExprGroups>& uniqueUses() {
+    return unique_uses_;
+  }
+
+  std::unordered_map<IdGroup, ExprGroups>& uniqueDefinitions() {
+    return unique_definitions_;
+  }
+
+  // Set id0 and id1 to mapped in disjointIdsSet[mode], attempt to propagate
+  // new mapping through id0/id1 definitions/uses.
+  void mapIds(IterDomain* id0, IterDomain* id1);
+
+  // Checks if expr0 and expr1 should map together, maps them together, and if
+  // expression propagation is on, propagates mapping through them. This should
+  // be the only call in IdGraph to mapThroughExpr
+  void maybeMapThroughExprs(Expr* expr0, Expr* expr1, bool forward);
+
+  // Map expr0 and expr1 with eachother, update unique_definitions_ unique_uses_
+  // TODO: Make this variant hidden?
+  void mapExprs(Expr* expr0, Expr* expr1);
+
+  // Checks if expr's are considered "the same" where sameness inputs and
+  // outputs in the same position across expressions map with  provided
+  // MappingMode. If the expressions are determined the same then
+  // if forward
+  //   will map outputs
+  // else
+  //   will map inputs
+  // in the provided mode.
+  // Returns if expressions were mapped through.
+  //
+  // TODO: Make this private
+  bool mapThroughExpr(Expr* first, Expr* second, bool forward);
+
+  // Map through loop swizzles, as input/output IterDomains are exact, only the
+  // order they're traversed differs.
+  void mapThroughLoopSwizzles();
+
+  // Maps iter domain pairs returned by calling that return mappings from
+  // IdGraph::isTrivialExpr on every expression in the graph.
+  void mapThroughTrivialExprs();
+
+  // Removes expressions from unique_definitions_ and unique_uses_ that return
+  // mappings from IdGraph::isTrivialExpr
+  void removeTrivialExprs();
```

Comment:
If trivial exprs are removed, will we get a DAG, or do we still have loops in the graph?